### PR TITLE
Add support for Multiple IDPs on Login screen

### PR DESCRIPTION
**Pull request by kaankabalak**

### CODE REVIEW

**Comment by harshavardhana** [comment]
Is there a screenshot or an animation to look at how this will work? @kaankabalak

**Comment by kaankabalak** [comment]
> Is there a screenshot or an animation to look at how this will work? @kaankabalak

Based on the identity providers added, the login screen will look as follows:

Screen Shot 2022-08-29 at 12 39 26 PM

**Comment by kaankabalak** [comment]
Here are the instructions to test this PR:

* Install Podman here: https://podman.io/getting-started/installation
* Create and start your Podman machine:
```
podman machine init
podman machine start
```
* Run `make podman-run` from https://github.com/minio/minio-iam-testing to start testing IDP services.
* Copy the models, portal-ui and restate directories from the `console` repository to `vendor/github.com/minio/console` located in the `minio` repository (I used this as a workaround because I only could get the SSO configuration to be seen from the Console embedded to the MinIO Server — would appreciate if anyone can point me in the right direction)
* Build the MinIO server binary and start your MinIO Server. Make sure that the server is started with alias `myminio` and Console on port 10000.
* Then, add identity providers as follows using `mc`:
```
mc admin idp set myminio openid dex1 config_url="http://localhost:5556/dex/.well-known/openid-configuration"     client_id="minio-client-app"     client_secret="minio-client-app-secret"     scopes="openid,groups"     redirect_uri="http://127.0.0.1:10000/oauth_callback"     display_name="Login via dex1"     role_policy="consoleAdmin"
```
```
mc admin idp set myminio openid dex2 config_url="http://localhost:5557/dex/.well-known/openid-configuration"     client_id="minio-client-app-2"     client_secret="minio-client-app-secret-2"     scopes="openid,groups"     redirect_uri="http://127.0.0.1:10000/oauth_callback"     display_name="Login via dex2"     role_policy="readwrite"
```
* Restart your server using `mc admin service restart myminio`
* Now, verify that you see multiple SSO login options when the Console is accessed on port 10000
* Additionally, make sure that the login screen is displayed properly for the cases where we don’t have any IDPs or if we have a single IDP

**Comment by harshavardhana** [comment]
> > Is there a screenshot or an animation to look at how this will work? @kaankabalak
> 
> Based on the identity providers added, the login screen will look as follows:
> 
> Screen Shot 2022-08-29 at 12 39 26 PM

Shouldn't this be a drop down selector? What if we have a  dozen IDPs to login from ?

This design seems to be not catering for such eventualities.

**Comment by oscarocastellanos** [comment]
> > > Is there a screenshot or an animation to look at how this will work? @kaankabalak
> > 
> > 
> > Based on the identity providers added, the login screen will look as follows:
> > Screen Shot 2022-08-29 at 12 39 26 PM
> 
> Shouldn't this be a drop down selector? What if we have a dozen IDPs to login from ?
> 
> This design seems to be not catering for such eventualities.

@kaankabalak On one of the early explorations when we adapted the SSO login I had something like this. Maybe this could work:
Captura de Pantalla 2022-08-29 a la(s) 10 20 36

Let me know what you think!

**Comment by kaankabalak** [comment]
> @kaankabalak On one of the early explorations when we adapted the SSO login I had something like this. Maybe this could work: Captura de Pantalla 2022-08-29 a la(s) 10 20 36
> 
> Let me know what you think!

@harshavardhana If this looks good, I can start working on this 👍

**Comment by harshavardhana** [comment]
> @harshavardhana If this looks good, I can start working on it +1

Yes it does look good to me 👍🏽

**Comment by kaankabalak** [comment]
@harshavardhana @oscarocastellanos, I have updated the design and the login screen for multiple IDPs now look as follows. Please let me know if any further changes are required. Thanks in advance!

Screen Shot 2022-08-31 at 1 06 04 AM

**Comment by oscarocastellanos** [comment]
> @harshavardhana @oscarocastellanos, I have updated the design and the login screen for multiple IDPs now look as follows. Please let me know if any further changes are required. Thanks in advance!
> 
> Screen Shot 2022-08-31 at 1 06 04 AM

Looks good Kaan 👍

**Comment by kaankabalak** [comment]
I have implemented the design feedback, I would appreciate it if you could take a look whenever you're available. Thanks in advance 👍 

@Alevsk @cniackz @bexsoft @jinapurapu @dvaldivia @reivaj05 @bexsoft

**Comment by bexsoft** [comment]
Review in progress

**Comment by bexsoft** [comment]
LGTM, just need to test the full redirect flow

**Review by bexsoft — CHANGES_REQUESTED** [review]
Cannot complete login flow using IDP with provided test steps / Auth0. After redirection I get the following error:

Screen Shot 2022-09-01 at 12 22 50

Trying with Auth0 / new steps provided in idp-test repo, I get an error similar to this:

Screen Shot 2022-09-01 at 14 52 29

**Comment by oscarocastellanos** [comment]
Agree with @dvaldivia. In case of the single IDP, the text in the drop down (now button) could be the IDP name and not the action.

**Comment by bexsoft** [comment]
> Cannot complete login flow using IDP with provided test steps / Auth0. After redirection I get the following error:
> 
> Screen Shot 2022-09-01 at 12 22 50
> 
> Trying with Auth0 / new steps provided in idp-test repo, I get an error similar to this:
> 
> Screen Shot 2022-09-01 at 14 52 29

Issue was related to my local environment. I can now login using this component.

**Review by bexsoft — APPROVED** [review]
LGTM

**Comment by kaankabalak** [comment]
> I think when there's a single IDP, we don't need the dropdown, just a button would suffice, it would also solve the redundancy of double `login with sso`

@dvaldivia @oscarocastellanos Done 👍

**Review by bexsoft — APPROVED** [review]
LGTM

**Review by dvaldivia — APPROVED** [review]

File: restapi/user_login.go [hunk 1]
```diff
@@ -149,23 +149,32 @@ func getLoginDetailsResponse(params authApi.LoginDetailParams, openIDProviders o
...
+	if len(openIDProviders) >= 1 {
 		loginStrategy = models.LoginDetailsLoginStrategyRedirect
-		// initialize new oauth2 client
-		oauth2Client, err := openIDProviders.NewOauth2ProviderClient(idpName, nil, r, GetConsoleHTTPClient())
-		if err != nil {
-			return nil, ErrorWithContext(ctx, err, ErrOauth2Provider)
+		for name, provider := range openIDProviders {
+			// initialize new oauth2 client
+			oauth2Client, err := openIDProviders.NewOauth2ProviderClient(name, nil, r, GetConsoleHTTPClient())
+			if err != nil {
```
Comment (bpedersen2) [inline restapi/user_login.go:161]: Should probaly not return, but skip to next provider in list.

Comment (harshavardhana) [reply restapi/user_login.go:161]: In-fact this code is terribly written the openIDProviders is a slice that implements NewOauth2ProviderClient however we are looping over it as well. 

I have no idea why it is written this way.

Comment (harshavardhana) [reply restapi/user_login.go:161]: https://github.com/minio/console/pull/3168 tries to clean this up and fix the problem as well.

**Pull request merged**